libraries/Microsoft.Bot.Builder.Core.Extensions/TimeoutWarningMiddleware.cs

```diff
@@ -0,0 +1,37 @@
+﻿using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Bot.Builder.Core.Extensions
+{
+    public class TimeoutWarningMiddleware : IMiddleware
+    {
+        /// <summary>
+        /// Period of time to wait before logging a warning to indicate a potential channel timeout before the bot responds. Defaults to 8000ms.
+        /// </summary>
+        private readonly int _timeoutPeriod;
+
+        public TimeoutWarningMiddleware(int timeoutPeriod = 8000)
+        {
+            _timeoutPeriod = timeoutPeriod;
+        }
+
+        public async Task OnProcessRequest(IBotContext context, MiddlewareSet.NextDelegate next)
+        {
+            Timer timeoutWarningTimer = new Timer(TimeoutWarningTimerCallback, context, _timeoutPeriod, 0);
```

Comment:
Pretty sure it's possible the `next()` handler can bubble up an `Exception` and right now the code doesn't handle that explicitly to ensure disposal of the timer. Suggest a `using()` block to handle disposal instead.

Reply:
Thanks for pointing this out. Updated.